File: 0117-populating-next-right-pointers-in-each-node-ii/0117-populating-next-right-pointers-in-each-node-ii.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if not root:
            return root
        
         # Starting at root level
        prev_level_root = root

        while True:
            next_level_head = None # Will point to first node of next level
            next_level_tail = None # Will keep linking nodes at next level

            # Traverse current level using .next pointers
            while prev_level_root:
                # Handle left child
                if prev_level_root.left is not None:
                    if next_level_tail is None:
                        next_level_head = prev_level_root.left
                        next_level_tail = prev_level_root.left
                    else:
                        next_level_tail.next = prev_level_root.left
                        next_level_tail = next_level_tail.next
                # Handle right child
                if prev_level_root.right is not None:
                    if next_level_tail is None:
                        next_level_head = prev_level_root.right
                        next_level_tail = prev_level_root.right
                    else:
                        next_level_tail.next = prev_level_root.right
                        next_level_tail = next_level_tail.next
                # Move sideways in current level
                prev_level_root = prev_level_root.next

            # No more levels to process
            if next_level_head is None:
                break

            # Traverse to next level this is to process the 
            # next left child and right child nodes
            prev_level_root = next_level_head
        
        return root

File: 0117-populating-next-right-pointers-in-each-node-ii/0117-populating-next-right-pointers-in-each-node-ii.js
/**
 * // Definition for a _Node.
 * function _Node(val, left, right, next) {
 *    this.val = val === undefined ? null : val;
 *    this.left = left === undefined ? null : left;
 *    this.right = right === undefined ? null : right;
 *    this.next = next === undefined ? null : next;
 * };
 */

/**
 * @param {_Node} root
 * @return {_Node}
 */
var connect = function(root) {
    // Logic: If parent node has the right child then
    // the parent's next should point to it;

    let currentNode = root;
    if(currentNode === null){
        return root;
    }

    let currentNodeIdx = 0;
    const queue = [];
    queue.push(currentNode);
    let totalNodes = 0;

    while(currentNodeIdx < queue.length){
        const levelLen = queue.length;
        totalNodes += (levelLen - currentNodeIdx);

        for(let idxI = currentNodeIdx; idxI < levelLen; idxI++){
            currentNode = queue[currentNodeIdx];
            currentNodeIdx += 1;

            let nextNode = null;
            if(currentNodeIdx < totalNodes){
                nextNode = queue[currentNodeIdx];
            }

            currentNode.next = nextNode;
            // console.log(currentNodeIdx, 'idx', totalNodes, queue.length);
            // console.log(currentNode, 'hello', nextNode, 'queue', queue[currentNodeIdx]);
            if(currentNode.left){
                queue.push(currentNode.left);
            }
            if(currentNode.right){
                queue.push(currentNode.right);
            }
        }
    }

    return root;
};